### PR TITLE
feat: add metadata field to API keys to aid organisational tracking and management

### LINKED ISSUE #1058: feat: add metadata field to API keys to aid organisational tracking and management
## Context

API keys in MaaS currently support only `name` and `description` fields for user-provided information. Organisations integrating MaaS with enterprise systems need to track additional structured metadata with API keys, such as:

- **Change Management Database (CMDB) IDs** to support enterprise service management needs such as "asset ownership" 
- **Cost center codes** for billing, chargeback, and budget tracking
- **Project identifiers** for organisational tracking across teams
- **Compliance levels** and security classifications for audit requirements
- **Owner contact information** (email, team) for operational communication
- **Custom organisational tags** for categorisation and reporting

Currently, the only option is to encode this information in the `description` field as JSON strings or delimited text. This approach has significant limitations:

- No type validation (invalid JSON accepted silently)
- Cannot query or filter by metadata fields (e.g., "find all keys for cost centre 123456")
- No indexing support for efficient searches
- Breaks the API contract (description is documented as human-readable text, not structured data)
- No schema enforcement or validation

Organisations need a first-class, structured metadata field that supports:
- Arbitrary key-value pairs
- Type-safe storage and retrieval
- Efficient queries and filtering
- Integration with enterprise systems (CMDB, billing, compliance tools)

## What needs to change

Add support for structured metadata on API keys, enabling organisations to store and query custom key-value pairs alongside the standard API key fields.

**Minimum requirements:**
1. API keys should support an optional metadata field for custom key-value storage
2. Metadata should be queryable (e.g., search for keys by CMDB ID or cost center)
3. Metadata should be type-safe (invalid data rejected)
4. Existing API keys without metadata should continue to work (backward compatibility)

**Example use cases:**

**Create key with CMDB tracking:**
```bash
curl -X POST "${MAAS_API_URL}/maas-api/v1/api-keys" \
  -H "Authorization: Bearer ${TOKEN}" \
  -d '{
    "name": "production-pipeline",
    "description": "Production data processing pipeline",
    "metadata": {
      "cmdb_id": "AST123456",
      "cost_center": "CC-DATA-001",
      "environment": "production",
      "project_code": "PROJ-ML-2024",
      "owner_email": "ml-team@company.com"
    }
  }'
```

**Search by CMDB ID:**
```bash
curl -X POST "${MAAS_API_URL}/maas-api/v1/api-keys/search" \
  -H "Authorization: Bearer ${TOKEN}" \
  -d '{
    "filters": {
      "metadataContains": {"cmdb_id": "AST123456"}
    }
  }'
```

**Search by cost center:**
```bash
curl -X POST "${MAAS_API_URL}/maas-api/v1/api-keys/search" \
  -H "Authorization: Bearer ${TOKEN}" \
  -d '{
    "filters": {
      "metadataContains": {"cost_center": "CC-DATA-001"}
    }
  }'
```

## Acceptance criteria

- API keys can be created with optional metadata field containing custom key-value pairs
- Metadata is returned when retrieving API keys (GET endpoint)
- Search endpoint supports filtering by metadata values
- Existing API keys without metadata continue to work unchanged (backward compatibility)
- Invalid metadata is rejected with clear error messages
- Documentation includes examples of metadata usage
- Tests cover metadata creation, retrieval, and search operations

## Additional context

**Why not use the description field?**
- Description is plain text, not structured data
- No query capability without full-text search
- Cannot enforce schemas or validate structure
- Would break existing API contract and user expectations

**Similar patterns in other systems:**
- Kubernetes uses `metadata.labels` and `metadata.annotations` for extensible key-value storage
- AWS uses resource tags for tracking and billing
- Cloud providers commonly support custom metadata on API keys and resources

**Benefits:**
- Enables integration with enterprise change management, billing, and compliance systems
- Supports organisational tracking and reporting requirements
- Provides flexibility without requiring schema changes for new metadata fields
- Follows industry-standard patterns for extensible metadata

### PR DESCRIPTION
**Pull request by bryonbaker**
## Description

Closes: #1058 

Added optional `labels` field to API keys, enabling organizations to attach custom key-value metadata for enterprise integration (CMDB tracking, cost center allocation, compliance, etc.).

**Storage**: PostgreSQL JSONB with GIN index for efficient queries  
**Backward Compatible**: NULL for existing keys, optional for new keys  

---

## Files Changed

- `maas-api/internal/api_keys/types.go`
  - `ApiKey` - Added `Labels map[string]string` field
  - `CreateAPIKeyRequest` - Added `Labels map[string]string` field
  - `SearchFilters` - Added `LabelsContain map[string]string` field
- `maas-api/internal/api_keys/handler.go`
  - `validateLabels()` function for label validation.
- `maas-api/internal/api_keys/service.go`
  - `CreateAPIKeyResponse` - Added `Labels map[string]string` field
  - `CreateAPIKey()` signature: added `labels map[string]string` parameter
  - `CreateAPIKeyResponse` struct: added `Labels` field
  - Response population: includes labels for user confirmation
  - Logging: added `"hasLabels", len(labels) > 0` for observability
- `maas-api/internal/api_keys/store_interface.go`
  - Updated persistence interface with labels parm.
- `maas-api/internal/api_keys/store_postgres.go`
  - Updated INSERT query to include `labels` column
  - Added `labels` to SELECT query
- `maas-api/internal/api_keys/store_mock.go`
  - `AddKey()`: added `labels` parameter, normalizes empty map to `nil`
  - `Search()`: added labels filtering logic
  - `labelsContain()` helper: mirrors PostgreSQL JSONB `@>` semantics
 
- Included regression updates plus new tests in the following test files:.
  - `test/e2e/test_api_keys.py` (4 new e2e tests)
  - `maas-api/internal/api_keys/handler_test.go`
  - `maas-api/internal/api_keys/service_test.go`
  - `maas-api/internal/api_keys/store_test.go`
  - `maas-api/internal/api_keys/store_postgres_test.go`

- Documentation changes:
  - `docs/content/user-guide/api-key-management.md` - guides and examples
  - `docs/content/contributing/testing-guide.md` - new integration test

---

## How Has This Been Tested?

Updated existing unit tests to include `nil` label to ensure no regressions were introduced.

**Verification:** All 59 existing tests pass - no regressions introduced.

Added new tests for labels: `maas-api/internal/api_keys/handler_test.go`

**Implemented tests:**

1. `TestValidateLabels_Success` - Valid label configurations (6 cases)
2. `TestValidateLabels_Errors` - Validation failures (10 cases)
3. `TestCreateAPIKey_WithLabels` - Creation with labels
4. `TestCreateAPIKey_WithoutLabels` - Backward compatibility
5. `TestCreateAPIKey_InvalidLabels` - Error handling (6 cases)
6. `TestSearchAPIKeys_ByLabels` - Search filtering (6 scenarios)
7. `TestGetAPIKey_WithLabels` - Retrieval with labels
8. `TestCreateEphemeralKey_InvalidLabels` - Error handling

**Result:** All 7 new tests passing.

---

### New Integration Tests

MockStore cannot calidate JSONB encoding and GIN index perforemance and NULL vs. empty object semantics so a three new integration tests were created. Test coverage: Validation scenarios, CRUD operations, search filtering, edge cases, and database-specific behaviour.

**File:** `maas-api/internal/api_keys/store_postgres_test.go`

1. `TestPostgresStore_LabelsRoundTrip` - JSONB marshaling/unmarshaling verification
2. `TestPostgresStore_SearchByLabels` - GIN index and `@>` containment operator
3. `TestPostgresStore_BackwardCompatibility_NullLabels` - NULL handling for legacy keys

Note: The process for running these tests is documented in the `./docs/content/contributing/testing-guide.md` document.

**Result:** All 3 integration tests passing.

---

### New E2E Tests

Added new E2E tests to `test/e2e/tests/test_api_keys.py` (new `TestAPIKeyLabels` class)

**Tests:**

1. `test_create_api_key_with_labels` - Create with labels, verify in response and GET
2. `test_search_api_keys_by_labels` - Search by `labelsContain` filter, verify JSONB containment
3. `test_labels_validation_errors` - Reject >50 labels, key >128 chars, invalid key characters
4. `test_backward_compatibility_no_labels` - Keys without labels return `null`/absent labels

**Run command:**

```bash
pytest ./test/e2e/tests/test_api_keys.py::TestAPIKeyLabels -v
```

### Test Coverage Summary

**Unit Tests:** 76 tests total

- 69 regression tests (existing functionality)
- 7 new tests (labels functionality)
- All passing 

**Integration Tests:** 3 tests (optional, requires PostgreSQL)

- All passing

**e2e Tests:** 4 tests total

```bash
test/e2e/tests/test_api_keys.py::TestAPIKeyLabels::test_create_api_key_with_labels PASSED          [25%]
test/e2e/tests/test_api_keys.py::TestAPIKeyLabels::test_search_api_keys_by_labels PASSED           [50%]
test/e2e/tests/test_api_keys.py::TestAPIKeyLabels::test_labels_validation_errors PASSED            [75%]
test/e2e/tests/test_api_keys.py::TestAPIKeyLabels::test_backward_compatibility_no_labels PASSED    [100%]
```

## Summary

* **New Features**
  * Added optional structured `labels` on API key creation; labels are validated, persisted, and returned in create and key details responses.
  * Added API key search filtering with `filters.labelsContain` (including AND behavior across multiple label pairs).
* **Bug Fixes**
  * Improved backward compatibility: keys created without labels return `labels` consistently as `null`/empty.
* **Documentation**
  * Updated API key management guide with label request/response examples, constraints, and label-based filtering examples.
  * Expanded the testing guide with PostgreSQL integration test instructions.
* **Database**
  * Added JSONB `labels` storage with indexing to support efficient label containment queries.
* **Tests**
  * Added unit, PostgreSQL integration, and end-to-end tests covering label persistence, validation errors, and label-based search.

### CODE REVIEW

**Comment by bryonbaker** [comment]
@coderabbitai review

**Comment by jland-redhat** [comment]
Hey Bryan, just wanted to ping back that our code freeze is happening on Friday so holding off on new functionality PRs until then.

But as soon as we are post code freeze we will prioritize getting this in, as it is valuable functionality.

**Comment by bryonbaker** [comment]
@jland-redhat - I just re-based with the latest upstream changes after code freeze and retested. There were a couple of merge conflicts things I needed to manage.  It is good to go and should make for a cleaner merge. Note the new integration test I created for database updates.

**Comment by jland-redhat** [comment]
Hey @bryonbaker 

We are past code freeze so good to look into getting this merged! Looks like a few test failed but seems to be a permissions issues that is not code related rerunning them

File: maas-api/db/schema/0006_add_labels_column.up.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+-- Add labels column as JSONB for structured key-value storage
+-- NULL for existing keys (backward compatibility)
+ALTER TABLE api_keys ADD COLUMN IF NOT EXISTS labels JSONB DEFAULT NULL;
+
+-- GIN index for efficient JSONB containment queries (@> operator)
+-- Supports fast lookups like: WHERE labels @> '{"cmdb_id": "AST123456"}'::jsonb
+CREATE INDEX IF NOT EXISTS idx_api_keys_labels_gin ON api_keys USING GIN (labels);
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply maas-api/db/schema/0006_add_labels_column.up.sql]: I agree with this comment. We should use `CREATE INDEX CONCURRENTLY` to avoid locking the whole table.

File: maas-api/db/schema/0006_add_labels_column.up.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+-- Add check constraint for labels structure (enforce object type)
+ALTER TABLE api_keys ADD CONSTRAINT api_keys_labels_is_object 
+    CHECK (labels IS NULL OR jsonb_typeof(labels) = 'object');
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply maas-api/db/schema/0006_add_labels_column.up.sql]: Valid concern called out. The check needs to include `IF NOT EXISTS`.

Comment (ishitasequeira) [inline maas-api/db/schema/0006_add_labels_column.up.sql]: The index creation can take longer. Can we move this constraint above index creation?

**Comment by bryonbaker** [comment]
Hi @jland-redhat 
I ran all the tests manually after rebasing. I noticed a test failure in unrelated code - looked like a failure of another change in unrelated code from 3.5. But all my tests passed. Also all regression tests related to api-key creation and retrieval passed. 

Update since edit: Found the source of the errors. I forgot to change go version to 1.25.8 for maas-api. Was still using 1.25.0 from maas-controller. Would be good to get these consiustent.

File: maas-api/internal/api_keys/types.go [hunk 3]
```diff
@@ -96,6 +97,9 @@ type SearchFilters struct {
 
 	// Ephemeral key filter
 	IncludeEphemeral *bool `json:"includeEphemeral,omitempty"` // Include ephemeral keys in results (default: false)
+
+	// Labels filter
+	LabelsContain  map[string]string `json:"labelsContain,omitempty"` // Filter by structured key-value pairs for API key metadata
```
[the remark this answers is not in the document]
Comment (ishitasequeira) [reply maas-api/internal/api_keys/types.go:102]: This comment is also worth addressing. Looks like we are adding validation on labels for creating an API key, but we do not have validations on `LabelContains` in Search call.

**Review by ishitasequeira — APPROVED** [review]
lgtm

**Review by jland-redhat — APPROVED** [review]
Lgtm
